File: python/LeetCode_29_DivideTwoIntegers.py
def divide(dividend: int, divisor: int) -> int:
    is_negative = 1 if (dividend > 0 and divisor < 0) or (dividend < 0 and divisor > 0) else 0
    dividend = abs(dividend)
    divisor = abs(divisor)

    result = 0
    while (dividend >= divisor):
        d = divisor
        mult = 1
        while (dividend >= d):
            dividend -= d 
            result += mult 
            mult += mult 
            d += d

    if (is_negative): result = -result
    return min(max(result, -2147483648), 2147483647)


divide(-2147483648, 1)
    

File: c/LeetCode_29_DivideTwoIntegers.c
#include <stdio.h>
#include <stdlib.h>


int divide(int dividend, int divisor) {
    int is_negative = (dividend > 0 && divisor < 0) || (dividend < 0 && divisor > 0) ? 1 : 0;
    dividend = abs(dividend);
    divisor = abs(divisor);

    int result = 0;
    while (dividend >= divisor) {
        int div = divisor;
        int mul = 1;
        while (dividend >= div) {
            dividend -= div;
            result += mul;
            mul += mul; 
            div += div;
        }
    }

    result = (is_negative) ? -result : result;
    
    if (result >= 2147483647) {
        return 2147483647;
    } else if (result <= -2147483648) {
        return -2147483648;
    }

    return result;
}


int main() {
    return 0;
}